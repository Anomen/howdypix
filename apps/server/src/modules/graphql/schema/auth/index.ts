export * from "./authTypes";
export * from "./authMutation";
