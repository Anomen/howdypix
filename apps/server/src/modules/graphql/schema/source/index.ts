export * from "./sourceTypes";
export * from "./sourceQuery";
