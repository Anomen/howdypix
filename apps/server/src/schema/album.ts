import { objectType } from "nexus";

export const Album = objectType({
  name: "Album",
  definition(t) {
    t.field("name", { type: "String" });
    t.field("dir", { type: "String" });
    t.field("source", { type: "String" });
  }
});
