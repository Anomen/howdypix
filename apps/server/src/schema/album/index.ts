export * from "./albumTypes";
export * from "./albumQuery";
