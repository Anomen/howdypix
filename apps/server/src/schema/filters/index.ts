export * from "./filtersTypes";
export * from "./filtersQuery";
