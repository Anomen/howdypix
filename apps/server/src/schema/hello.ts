import { objectType } from "nexus";

export const Hello = objectType({
  name: "Hello",
  definition(t) {
    t.id("id");
    t.string("toto", { nullable: true });
    t.string("type", { nullable: true });
  }
});
