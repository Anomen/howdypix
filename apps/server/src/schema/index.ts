export * from "./album";
export * from "./auth";
export * from "./user";
