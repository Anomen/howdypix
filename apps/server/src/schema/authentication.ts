import { enumType, mutationField, objectType, stringArg } from "nexus";
import { createTransport } from "nodemailer";
import smtpTransport from "nodemailer-smtp-transport";
import { state } from "../state";
import { find } from "lodash";
import { NexusGenFieldTypes } from "@howdypix/graphql-schema/schema";
import { magickLink } from "../email";

export const SendEmailMessage = enumType({
  name: "SendEmailMessage",
  members: ["AUTH_EMAIL_OK", "AUTH_EMAIL_ERR_NOT_EXIST", "AUTH_EMAIL_ERR"],
  description:
    "The type of message that the user can get when requesting a magic link."
});

export const SendEmailType = objectType({
  name: "SendEmailType",
  definition(t) {
    t.field("messageId", { type: "SendEmailMessage" });
  }
});

export const sendEmail = mutationField("sendEmail", {
  type: "SendEmailType",
  args: {
    email: stringArg()
  },
  resolve: async (root, args) =>
    new Promise<NexusGenFieldTypes["Mutation"]["sendEmail"]>(resolve => {
      const user = find(state.userConfig.users, u => u.email == args.email);
      console.log(user);

      if (user) {
        const transporter = createTransport(
          smtpTransport({
            host: process.env.SMTP_HOST,
            port: parseInt(process.env.SMTP_PORT as string),
            ignoreTLS: true
          })
        );

        transporter.sendMail(
          {
            from: `${state.userConfig.emailSender.name}<${state.userConfig.emailSender.email}>`,
            to: `${user.name}<${user.email}>`,
            html: magickLink({ code: "123", name: user.name })
          },
          (error, info) => {
            if (error) {
              resolve({
                messageId: "AUTH_EMAIL_ERR"
              });
            } else {
              resolve({
                messageId: "AUTH_EMAIL_OK"
              });
            }
          }
        );
      } else {
        resolve({
          messageId: "AUTH_EMAIL_ERR_NOT_EXIST"
        });
      }
    })
});
