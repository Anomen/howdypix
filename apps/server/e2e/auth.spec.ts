// Load env variables
import { NexusGenFieldTypes } from "@howdypix/graphql-schema/schema";

require("dotenv").config();

import { client, fetchTokens } from "./utils";
import gql from "graphql-tag";
import axios, { AxiosError } from "axios";

import { routes } from "@howdypix/utils";
import { UserInfo } from "@howdypix/shared-types";

describe("The authentication routes", () => {
  test("should reject if the code is not valid", async () => {
    const response = await axios.post(routes.codeValidation.value());
    expect(response.data).toMatchSnapshot();
  });

  describe("validating the email", function () {
    const sendMutation = async (email: string) =>
      client.mutate({
        mutation: gql`
          mutation AuthEmail($email: String!) {
            authEmail(email: $email) {
              messageId
              messageData
            }
          }
        `,
        variables: {
          email,
        },
      });

    test("should return an error if the email is not allowed", async () => {
      try {
        expect(await sendMutation("NONE")).toMatchSnapshot();
      } catch (e) {
        console.log(e.networkError.result);
      }
    });

    test("should return a success message if the email is allowed", async () => {
      expect(await sendMutation("dev@vickev.com")).toMatchSnapshot();
    });
  });

  describe("validating the magic link", function () {
    test("should return tokens if the magic link is correct", async () => {
      const data = await fetchTokens();

      expect(data).toHaveProperty("token");
      expect(data).toHaveProperty("refreshToken");
      expect(data.user).toMatchSnapshot();
    });

    test("should have an error if the link is not correct", async () => {
      const data = await fetchTokens("wrong-code");
      expect(data).toMatchSnapshot();
    });
  });

  describe("fetching the user info", function () {
    test("should return the user info if the token is correct", async () => {
      const { token, user } = await fetchTokens();

      const response = await axios.post<UserInfo>(
        routes.authenticatedUser.value(),
        {
          token,
        }
      );

      expect(response.data).toEqual(user);
    });

    test("should return an error if the token is wrong", async () => {
      try {
        const response = await axios.post<UserInfo>(
          routes.authenticatedUser.value(),
          {
            token: "wrong",
          }
        );
      } catch (e) {
        expect((e as AxiosError).response.status).toEqual(401);
      }
    });
  });

  describe("refreshing the token", function () {
    test("should return a new token", async (done) => {
      const { refreshToken, token } = await fetchTokens();

      setTimeout(async () => {
        const response = await axios.post<{ token: string }>(
          routes.refreshToken.value(),
          {
            token: refreshToken,
          }
        );

        expect(response.data.token).toBeDefined();
        expect(response.data.token).not.toEqual(token);

        done();
      }, 1000);
    });

    test("should return an error if the token is wrong", async () => {
      try {
        const response = await axios.post<{ token: string }>(
          routes.refreshToken.value(),
          {
            token: "wrong",
          }
        );
      } catch (e) {
        expect((e as AxiosError).response.status).toEqual(401);
      }
    });
  });
});
