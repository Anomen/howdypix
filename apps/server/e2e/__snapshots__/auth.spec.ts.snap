// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The authentication routes should reject if the code is not valid 1`] = `
Object {
  "error": "Token expired",
}
`;

exports[`The authentication routes validating the email should return a success message if the email is allowed 1`] = `
Object {
  "data": Object {
    "authEmail": Object {
      "__typename": "AuthEmailType",
      "messageData": null,
      "messageId": "AUTH_EMAIL_OK",
    },
  },
}
`;

exports[`The authentication routes validating the email should return an error if the email is not allowed 1`] = `
Object {
  "data": Object {
    "authEmail": Object {
      "__typename": "AuthEmailType",
      "messageData": null,
      "messageId": "AUTH_EMAIL_ERR_NOT_EXIST",
    },
  },
}
`;

exports[`The authentication routes validating the magic link should have an error if the link is not correct 1`] = `
Object {
  "error": "Token expired",
}
`;

exports[`The authentication routes validating the magic link should return tokens if the magic link is correct 1`] = `
Object {
  "email": "dev@vickev.com",
}
`;
